(dev/core#5614) Search Kit - Allow DB Entity with custom column names

totten opened the pull request:
Overview
----------------------------------------

When creating a saved-search, one may define a "DB Entity" (eg SQL TABLE or VIEW derived from the query). This update allows the administrator to define the SQL column names.

(See also: https://lab.civicrm.org/dev/core/-/issues/5614)

Steps to Reproduce
----------------------------------------

* Create a new saved-search
* Choose some columns and criteria
* Add a search-display of type "DB Entity"
* Configure the list of columns for the entity

Before
----------------------------------------

For each column, you may define the visible label:

<img width="707" alt="Screenshot 2025-01-31 at 10 44 41 PM" src="https://github.com/user-attachments/assets/cf1b374e-b2e5-4d20-aded-fc462deb5772" />

These labels will appear as the headings in downstream searches.

The SQL column-name is auto-generated from the original `SELECT` expression that provided the data. In simple scenarios, the generated name is intuitive. But in complex scenarios, the name becomes opaque (eg `GROUP_CONCAT_Contact_RelationshipCache_Conc51a580ccf5b246a`).

After
----------------------------------------

For each column, you may define both the SQL column-name and the visible label.

<img width="698" alt="Screenshot 2025-01-31 at 10 43 20 PM" src="https://github.com/user-attachments/assets/0715278e-75a0-4786-b304-75db8f2fdfae" />

Technical Details
----------------------------------------

* Existing configurations for DB entities do not have an explicit `name` property -- they strictly rely on auto-generated column-names. For continuity, this data is treated the same as before. (*A blank `name` means that the column-name will be auto-generated.*)
* Additionally, this adds some guards to the admin UI -- if you try to rename an existing column, it will show warning like this:

    <img width="574" alt="Screenshot 2025-01-31 at 10 53 53 PM" src="https://github.com/user-attachments/assets/3f4e9c5a-be04-4d60-8604-40dac7d92f08" />

    Similarly, if you try to delete a DB Entity, it will also show a warning:

    <img width="580" alt="Screenshot 2025-01-31 at 10 54 18 PM" src="https://github.com/user-attachments/assets/e894f7e0-6df3-4370-b843-7af2bdccc3e7" />

totten committed:
(dev/core#5614) SK UI - Accept a custom "name" for each column in a custom entity

totten committed:
(dev/core#5614) SK - When generating custom DB Entity, respect assigned "name"s

totten committed:
crm-confirm - Accept option "confirmed:true"

The value of `confirmed` can be pre-computed.  If set to true, then we can
skip the confirmation dialog.

totten committed:
crmSearchAdmin - If saving critical changes, prompt to confirm

totten committed:
(REF) crmSearchAdmin - Extract internal helper, fireHooks()

totten committed:
searchAdminDisplayEntity - Warn if external references to DB entity may become stale

colemanw commented on ext/search_kit/ang/crmSearchAdmin/crmSearchAdmin.html:
```diff
@@ -18,7 +18,8 @@ <h1 crm-page-title>{{ ($ctrl.savedSearch.is_template ? ts('Search Template') : t
       <div class="crm-flex-1 form-inline">
         <div class="btn-group">
-          <button type="button" class="btn" ng-class="{'btn-primary': status === 'unsaved', 'btn-warning': status === 'saving', 'btn-success': status === 'saved'}" ng-disabled="status !== 'unsaved'" ng-click="$ctrl.save()">
+          <div crm-ui-debug="$ctrl"></div>
```
Stray debug statement?

totten replied:
Doh. Cleaned+squashed.

colemanw commented on ext/search_kit/ang/crmSearchAdmin/crmSearchAdmin.component.js:
```diff
@@ -156,6 +167,30 @@
+    // Generate the confirmation dialog
+    this.confirmSave = function() {
+      const targets = {}, data = {messages: []};
+      var newCount = 0;
```
FYI I've been trying to get into the habit of using `let` instead of `var` when writing new code.

totten replied:
Good point. `let` does sound more intuitive.

herbdool commented:
I haven't tested but this appears to be exactly what I was imagining plus more.

totten commented:
> * Existing configurations for DB entities do not have an explicit `name` property -- they strictly rely on auto-generated column-names. For continuity, this data is treated the same as before. (*A blank `name` means that the column-name will be auto-generated.*)

I still think this is good in terms of handling upgrades safely.

However, there is a visual factor that would feel quirky (esp over time) -- in the UI, it defaults to an empty `name` for each column.  It would feel more natural if the `name` UI was visibly prepopulated. And when doing prepopulation, there are couple a ways to go (and I'm not certain difficulty of either):

* __New-columns only__: Only prepopulate the `name` for new columns. Any existing columns remain as `name=""` (implying the use of the current/legacy formula).
    * On the plus side -- when iterating on the formula in the future, there's little risk. The formula only define a _default suggestion_. It has no consequences for interoperability or upgrades.
    * On the down side --  existing columns still look quirky (`name=""`), and you may end-up with two formulas.(The "new column" formula vs the "legacy/upgrade" formula.)
* __New-columns and pre-existing columns__: Take the existing formula. Extend it to be apparent in the UI for any/all columns.
    * On the plus side -- this applies equally to new and pre-existing columns. There is only one formula.
    * On the down side -- Changing the formula could cause surprise schema-changes in future upgrades. (Between version#s, circumstantial visits to the admin UI, and the use mgd's for SK.... it's hard to plot all the upgrade-paths.) The simple/effective thing is to lock-down the meaning of `name=""` to match the current formula)

(In both cases, there's probably some consideration about how the JS+PHP code interact, though I'm not sure which variations are easy/hard.)

totten commented on ext/search_kit/ang/crmSearchAdmin/crmSearchAdmin.html:
```diff
@@ -18,7 +18,7 @@ <h1 crm-page-title>{{ ($ctrl.savedSearch.is_template ? ts('Search Template') : t
       <div class="crm-flex-1 form-inline">
         <div class="btn-group">
-          <button type="button" class="btn" ng-class="{'btn-primary': status === 'unsaved', 'btn-warning': status === 'saving', 'btn-success': status === 'saved'}" ng-disabled="status !== 'unsaved'" ng-click="$ctrl.save()">
+          <button type="button" class="btn" ng-class="{'btn-primary': status === 'unsaved', 'btn-warning': status === 'saving', 'btn-success': status === 'saved'}" ng-disabled="status !== 'unsaved'" crm-confirm="$ctrl.confirmSave()" on-yes="$ctrl.save()">
```
Small thought: There are common critiques of confirmation dialogs -- they get annoying; people get de-sensitized; etc. That's well and good. 

IMHO, this is not that:

* The call to `$ctrl.confirmSave()` is a computation that decides whether a prompt is worthwhile.
* For most usual edits to regular saved-searches, it will proceed as directed by the user. No need to confirm.
* It only prompts for confirmation if you are deleting or renaming a schema-element (because other things might depend on the schema-element). And it specifically identifies the changed elements.

colemanw commented:
> @totten  It would feel more natural if the name UI was visibly prepopulated. And when doing prepopulation, there are couple a ways to go (and I'm not certain difficulty of either):

I would vote for option 3: use the formula to set the `placeholder` attribute of the field. That way it can be left blank and the user will know the default value. This will apply equally well to new and existing displays.

totten committed:
(JS) Add port of CRM.utils.munge()

totten committed:
(JS) Add md5 helper

totten committed:
DB Entity - Display calculated column "name" as an HTML placeholder

totten committed:
composer.lock - Update checksum

totten committed:
(REF) Extract method createSqlName(). Add test.

totten commented:
> (@colemanw) I would vote for option 3: use the formula to set the placeholder attribute of the field. That way it can be left blank and the user will know the default value. This will apply equally well to new and existing displays.

Yeah, that sounds pretty sensible for UI. Pushed up a few commits to make it work. Here's how it looks with defaults on the first three columns (*with custom value on the fourth*):

![Screenshot from 2025-02-18 22-37-31](https://github.com/user-attachments/assets/873a89d7-3bd1-44e6-94f1-b3428842b194)

It took some finagling. When adding new entities/displays/fields, we don't yet know the server's opinion about the names -- and sending speculative AJAX requests feels heavy handed. It's lighter-touch to do that on client-side. So I re-implemented on client-side and hardened the convention a bit more (i.e. adding some tests and cross-references between PHP<=>JS code).

totten committed:
DB Entity - When user customizes column-name, conform to naming limits

These match the guidelines from Civi\Search\Meta::createSqlName()
and searchMeta.createSqlName(), i.e.

* Names should be 58 chars or fewer. (Longer names trigger
  overflow/hashing/rewriting.)

* Names should be limited to alphanumerics and underscore. (Other
  chars trigger an exception or rewrite.)

totten commented on ext/search_kit/ang/crmSearchAdmin/displays/searchAdminDisplayEntity.html:
```diff
@@ -56,6 +56,8 @@
       </legend>
       <div class="form-inline crm-search-admin-flex-row">
+        <label for="crm-search-admin-edit-col-{{ $index }}-name">{{:: ts('Name') }}</label>
+        <input id="crm-search-admin-edit-col-{{ $index }}-name" class="form-control crm-flex-1" type="text" ng-model="col.name" placeholder="{{$ctrl.createSqlName(col.key)}}">
```
Yeah, probably. The server behavior has some guards, but the user will be much happier (less confused) if we keep the name within the bounds of 58-chars and alphanumeric/underscore. I've added guards for both parts.

colemanw commented:
@totten probably woulda been fine to just cut off the placeholder at 58 chars with an ellipsis since a longer string probably won't fit in that textfield anyway, but... cool md5 for js!

totten commented:
Yeah, getting md5 in this context was more troublesome than you'd expect. (*I had to look through ~6 implementations before finding one with sufficient compatibility, minimal dependencies, and real version numbers.*) But happily md5 is common enough that we could find an option that slots-in.

Anyway, big picture, I didn't mind leaning into the obscurity -- because the outcome is that it actually makes the formula less important. Now the user can promptly see (and fix!) any quirky names.
